### PR TITLE
Fixed #34496 -- Fixed handling source maps with data URI in ManifestStaticFilesStorage.

### PR DESCRIPTION
**Pull request by hwalinga**
Fixes https://code.djangoproject.com/ticket/34496

The PR also includes an example as a test case.

### CODE REVIEW

File: tests/staticfiles_tests/project/documents/cached/module.js [hunk 1]
```diff
@@ -20,3 +20,4 @@ export {
     firstVar as firstVarAlias,
     secondVar as secondVarAlias
 } from "./module_test.js";
+//# sourceMappingURL=data:application/json;charset=utf8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIl9zcmMv
```
Comment (felixxm) [inline tests/staticfiles_tests/project/documents/cached/module.js:23]: Please move it to the `source_map.js`.

File: tests/staticfiles_tests/test_storage.py [hunk 2]
```diff
@@ -661,6 +661,14 @@ def test_aggregating_modules(self):
                 with self.subTest(module_import=module_import):
                     self.assertIn(module_import, content)
 
+    def test_data_uri_source_mapping(self):
+        relpath = self.hashed_file_path("cached/module.js")
+        self.assertEqual(relpath, "cached/module.242b85cc6818.js")
+        with storage.staticfiles_storage.open(relpath) as relfile:
+            content = relfile.read()
+            test = b"//# sourceMappingURL=data:application/json;charset=utf8;base64,eyJ"
```
Comment (felixxm) [inline tests/staticfiles_tests/test_storage.py:669]:
```suggestion
            test = (
                b"//# sourceMappingURL=data:application/json;charset=utf8;base64,"
                b"eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIl9zcmMv"
            ) 
```

File: django/contrib/staticfiles/storage.py [hunk 3]
```diff
@@ -86,7 +86,7 @@ class HashedFilesMixin:
                 ),
                 (
                     (
-                        r"(?m)(?P<matched>)^(/\*#[ \t]"
+                        r"(?m)^(?P<matched>/\*#[ \t]"
```
Comment (felixxm) [inline django/contrib/staticfiles/storage.py:89]: This change is untested. Add an example with data URI to the `source_map.css`.

**Review by felixxm — COMMENTED** [review]
@hwalinga Thanks :+1:

File: tests/staticfiles_tests/test_storage.py (hunk 2 above)
Comment (hwalinga) [reply tests/staticfiles_tests/test_storage.py]: I don't really like implicit string joining in Python. So I have added this with a `+` between.

File: django/contrib/staticfiles/storage.py (hunk 3 above)
Comment (hwalinga) [reply django/contrib/staticfiles/storage.py:89]: Done

File: tests/staticfiles_tests/project/documents/cached/module.js (hunk 1 above)
Comment (hwalinga) [reply tests/staticfiles_tests/project/documents/cached/module.js]: Made it a separate test named test_js_source_map_uri with a source_map_data_uri.js and the same for CSS.

File: tests/staticfiles_tests/test_storage.py (hunk 2 above)
Comment (felixxm) [reply tests/staticfiles_tests/test_storage.py]: > I don't really like implicit string joining in Python. So I have added this with a `+` between.

I understand, but we cannot follow contributor personal preferences, please remove "+" and update as requested.

**Comment by felixxm** [comment]
@hwalinga Thanks :+1: Welcome aboard :boat:

**Comment by hwalinga** [comment]
@felixxm Great!

As this is actually a software regression (this bug was not present in 3.2 and is in 4.2), should it also be backported to the 4.2 branch?

**Comment by felixxm** [comment]
> As this is actually a software regression (this bug was not present in 3.2 and is in 4.2), should it also be backported to the 4.2 branch?

This is a regression in Django 4.0, but it was reported when Django 4.0 was no longer supported, per our backporting policy this means it doesn't qualify for a backport to 4.0.x, 4.1.x, or 4.2.x anymore. See [Django’s release process](https://docs.djangoproject.com/en/stable/internals/release-process/) for more details.

**Pull request merged**